lib/AST/Expr.cpp

```diff
@@ -1068,6 +1068,22 @@ Type OverloadSetRefExpr::getBaseType() const {
   llvm_unreachable("Unhandled overloaded set reference expression");
 }
 
+bool ApplyExpr::shouldApplyDistributedThunk() const {
+  // only a distributed decl has a chance of needing to be invoked as a thunk
+  auto func = dyn_cast<AbstractFunctionDecl>(getCalledValue());
+  if (!func || !func->isDistributed())
+    return false;
+
+  if (implicitlyThrows())
+    return true;
+
+  auto isEffectivelyAsync = func->hasAsync() || isImplicitlyAsync();
+  if (func->hasThrows() && isEffectivelyAsync)
+    return true;
```

Comment:
I don't think this is correct. You can have something like this:

```swift
distributed actor A {
  distributed func f() async throws { }
  func g() {
    Task {
       try await f() // shouldn't use the distributed thunk!
     }
  }
}
```

Only the type checker can determine whether we're using the distributed thunk or not, and we'll need a bit here in `ApplyExpr` to keep track of that.

Reply:
Hm ah right... I'll figure that out and add a bit